src/spawn.c

```diff
@@ -0,0 +1,1121 @@
+	startup.wShowWindow = SW_SHOWMINNOACTIVE;  /* so you can switch to it if needed */
+
+	if (stdin_fd || stdout_fd || stderr_fd)
+	{
+		/* Not all programs accept mixed NULL and non-NULL hStd*, so we create all */
+		startup.dwFlags |= STARTF_USESTDHANDLES;
+
+		for (i = 0; i < 3; i++)
+		{
+			static int pindex[3][2] = { { READ_STDIN, WRITE_STDIN },
+					{ READ_STDOUT, WRITE_STDOUT }, { READ_STDERR, WRITE_STDERR } };
+			DWORD state;
+
+			if (!CreatePipe(&hpipe[pindex[i][0]], &hpipe[pindex[i][1]], NULL, 0))
+			{
+				hpipe[pindex[i][0]] = hpipe[pindex[i][1]] = NULL;
+				failed = "CreatePipe";
+				goto leave;
+			}
+
+			if (fd[i])
+			{
+				static int mode[3] = { _O_WRONLY, _O_RDONLY, _O_RDONLY };
+
+				if (!GetNamedPipeHandleState(hpipe[i], &state, NULL, NULL, NULL, NULL, 0))
+				{
+					failed = "GetNamedPipeHandleState";
+					goto leave;
+				}
+
+				state |= PIPE_NOWAIT;
+				if (!SetNamedPipeHandleState(hpipe[i], &state, NULL, NULL))
+				{
+					failed = "SetNamedPipeHandleState";
+					goto leave;
+				}
+
+				if ((*fd[i] = _open_osfhandle((intptr_t) hpipe[i], mode[i])) == -1)
+				{
+					failed = "_open_osfhandle";
+					message = g_strdup(g_strerror(errno));
+					goto leave;
+				}
+			}
+			else if (!CloseHandle(hpipe[i]))
+			{
+				failed = "CloseHandle";
+				goto leave;
+			}
+
+			if (!SetHandleInformation(hpipe[i + 3], HANDLE_FLAG_INHERIT,
+				HANDLE_FLAG_INHERIT))
+			{
+				failed = "SetHandleInformation";
+				goto leave;
+			}
+		}
+	}
+
+	startup.hStdInput = hpipe[READ_STDIN];
+	startup.hStdOutput = hpipe[WRITE_STDOUT];
+	startup.hStdError = hpipe[WRITE_STDERR];
+
+	if (!CreateProcess(NULL, command_line, NULL, NULL, TRUE, NORMAL_PRIORITY_CLASS,
+		environment, working_directory, &startup, &process))
+	{
+		failed = "CreateProcess";
+		/* no errors beyond this point */
+	}
+	else
+	{
+		failed = NULL;
+		*hprocess = process.hProcess;
+	}
+
+leave:
+	if (failed)
+	{
+		if (!message)
+			message = g_win32_error_message(GetLastError());
+
+		failure = g_strdup_printf("%s() failed: %s", failed, message);
+		g_free(message);
+	}
+
+	if (startup.dwFlags & STARTF_USESTDHANDLES)
+	{
+		for (i = 0; i < 3; i++)
+		{
+			if (failed)
+			{
+				if (fd[i] && *fd[i] != -1)
+					_close(*fd[i]);
+				else if (hpipe[i])
+					CloseHandle(hpipe[i]);
+			}
+
+			if (hpipe[i + 3])
+				CloseHandle(hpipe[i + 3]);
+		}
+	}
+
+	LeaveCriticalSection(&critical);
+	return failure;
+}
+
+
+static void spawn_append_argument(GString *command, const char *text)
+{
+	const char *s;
+
+	g_string_append_c(command, ' ');
+	for (s = text; *s; s++)
+		if (*s == '"' || g_ascii_isspace(*s))  /* are all RTL-s standartized? */
+			break;
+
+	if (*text && !*s)
+		g_string_append(command, text);
+	else
+	{
+		g_string_append_c(command, '"');
+
+		for (s = text; *s; s++)
+		{
+			const char *slash;
+
+			for (slash = s; *slash == '\\'; slash++);
+
+			if (slash > s)
+			{
+				g_string_append_len(command, s, slash - s);
+
+				if (strchr("\"", *slash))

```

Comment:
BTW, I don't get the logic here.  IIUC, if the `\`es aren't followed by `"` or at the end of the string, you don't escape them?  is that really what Windows want??